lib/api/apiUtils/object/validateChecksumHeaders.js

```diff
@@ -0,0 +1,32 @@
+const { errors } = require('arsenal');
+
+const { possibleSignatureChecksums, supportedSignatureChecksums } = require('../../../../constants');
+
+function validateChecksumHeaders(headers) {
+    // If the x-amz-trailer header is present the request is using one of the
+    // trailing checksum algorithms, which are not supported.
+    if (headers['x-amz-trailer'] !== undefined) {
+        return errors.BadRequest.customizeDescription('trailing checksum is not supported');
+    }
+
+    const signatureChecksum = headers['x-amz-content-sha256'];
+    if (signatureChecksum === undefined) {
+        return null;
+    }
+
+    if (supportedSignatureChecksums.has(signatureChecksum)) {
+        return null;
+    }
+
+    // If the value is not one of the possible checksum algorithms
+    // the only other valid value is the actual sha256 checksum of the payload.
+    // Do a simple sanity check of the length to guard against future algos.
+    // If the value is an unknown algo, then it will fail checksum validation.
+    if (!possibleSignatureChecksums.has(signatureChecksum) && signatureChecksum.length === 64) {
```

Comment:
I am surprised that the checksum should be 64 characters in length, because that would imply that it is always passed in hexadecimal encoding, but the documentation states that it should be in base64, this is what it says about the header `x-amz-checksum-sha256`:

> This header can be used as a data integrity check to verify that the data received is the same data that was originally sent. This header specifies the base64-encoded, 256-bit SHA-256 digest of the object. For more information, see Checking object integrity in the Amazon S3 User Guide.

From https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html

Reply:
That's a different header.
The API docs doesn't list `x-amz-content-sha256` on the individual operations page for some reason, but it's vaguely documented over here https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
It defines it as "Actual payload checksum value"
From observation it seems to be hex encoded.